src/lua/protobuf.lua

```diff
@@ -657,9 +726,142 @@ end
 
 -- }}} Encoders
 
+-- {{{ Decoders
+
+local function read_unknown_field(ctx, type)
+    local res = string.sub(ctx.val, ctx.pos - 1, ctx.pos - 1)
+    local begin = ctx.pos
+    if type == WIRE_TYPE_LEN then
+        local len = wireformat.decode_int(ctx)
+        ctx.pos = ctx.pos + len
+    elseif type == WIRE_TYPE_I32 then
+        ctx.pos = ctx.pos + 4
+    elseif type == WIRE_TYPE_I64 then
+        ctx.pos = ctx.pos + 8
+    else
+        local _ = wireformat.decode_int(ctx)
+    end
+    res = res .. string.sub(ctx.val, begin, ctx.pos - 1)
+    return res
+end
+
+local function decode_repeated(protocol, field_def, ctx)
+    local res = {}
+    if ctx.packed then
+        local len = wireformat.decode_int(ctx)
+        local start_pos = ctx.pos
+        local value
+        while ctx.pos < start_pos + len do
+            ctx.ignore_repeated = true
+            value = decode_field(protocol, field_def, ctx)
+            table.insert(res, value)
+        end
+    else
+        repeat
+            local field_id
+        ctx.ignore_repeated = true
+            local value = decode_field(protocol, field_def, ctx)
+            table.insert(res, value)
+            if ctx.pos > string.len(ctx.val) then
+                return res
+            else
+                local _, res_id = wireformat.decode_tag(ctx)
+                field_id = res_id
+            end
+        until field_id ~= field_def.id
+        ctx.pos = ctx.pos - 1
+    end
+    return res
+end
+
+decode_field = function(protocol, field_def, ctx)
+    local res
+    if field_def.repeated and not ctx.ignore_repeated then
+        res = decode_repeated(protocol, field_def, ctx)
+    elseif is_scalar(field_def) then
+        local scalar_def = scalars[field_def.type]
+        res = scalar_def.decode(ctx)
+        validate_scalar(field_def, res)
+    elseif is_enum(protocol, field_def) then
+        local id = wireformat.decode_int(ctx)
+        local curr_enum = protocol[field_def.type]
+        if curr_enum.value_by_id[id] then
+            res = curr_enum.value_by_id[id]
+        else
+            res = id
+        end
+    elseif is_message(protocol, field_def) then
+        local msg_data = wireformat.decode_len(ctx)
+        res = decode(protocol, field_def.type, msg_data, ctx.opts)
+    end
+    return res
+end
+
+decode = function(protocol, message_name, data, opts)
+    local opts = opts or {}
+    assert(type(opts) == 'table')
+    local res = {}
+    local ctx = {val = data, pos = 1, opts = opts}
```

Comment:
Let's describe all the fields of this context somewhere.

I suggest to create a constructor function for the decoding context and set unneded fields to `nil` explicitly. This way we can comment them in the same way as we do for non-nil fields.

```lua
local function new_decode_context(data, decode_opts)
    return {
        -- The source data to decode.
        val = data,
        -- The next position in the data to read.
        pos = 1,
        -- Blah blah.
        foo = nil,
    }
end
```

Reply:
OK, but now it is called `form_context` that feels unusual. How about `new_decode_context`?

Reply:
Renamed function with `new_decode_context`.